executor/aggfuncs/func_count_distinct.go

```diff
@@ -433,3 +435,387 @@ func appendJSON(encodedBytes, _ []byte, val json.BinaryJSON) []byte {
+// has power law distribution (i.e. in cases when most of the data sets are small).
+// This algorithm is also very accurate for data sets with small cardinality and very efficient on CPU. If number of
+// distinct element is more than 2^32, relative error may be high.
+type partialResult4ApproxCountDistinct struct {
+	size       uint32 /// Number of elements.
+	sizeDegree uint8  /// The size of the table as a power of 2.
+	skipDegree uint8  /// Skip elements not divisible by 2 ^ skipDegree.
+	hasZero    bool   /// The hash table contains an element with a hash value of 0.
+	buf        []approxCountDistinctHashValue
+}
+
+// NewPartialResult4ApproxCountDistinct builds a partial result for agg function ApproxCountDistinct.
+func NewPartialResult4ApproxCountDistinct() *partialResult4ApproxCountDistinct {
+	p := &partialResult4ApproxCountDistinct{}
+	p.reset()
+	return p
+}
+
+func (p *partialResult4ApproxCountDistinct) InsertHash64(x uint64) {
+	// no need to rehash, just cast into uint32
+	p.insertHash(approxCountDistinctHashValue(x))
+}
+
+func (p *partialResult4ApproxCountDistinct) alloc(newSizeDegree uint8) {
+	p.size = 0
+	p.skipDegree = 0
+	p.hasZero = false
+	p.buf = make([]approxCountDistinctHashValue, uint32(1)<<newSizeDegree)
+	p.sizeDegree = newSizeDegree
+}
+
+func (p *partialResult4ApproxCountDistinct) reset() {
+	p.alloc(uniquesHashSetInitialSizeDegree)
+}
+
+func max(a, b uint8) uint8 {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+func (p *partialResult4ApproxCountDistinct) bufSize() uint32 {
+	return uint32(1) << p.sizeDegree
+}
+
+func (p *partialResult4ApproxCountDistinct) mask() uint32 {
+	return p.bufSize() - 1
+}
+
+func (p *partialResult4ApproxCountDistinct) place(x approxCountDistinctHashValue) uint32 {
+	return uint32(x>>uniquesHashBitsForSkip) & p.mask()
+}
+
+// Increase the size of the buffer 2 times or up to new size degree.
+func (p *partialResult4ApproxCountDistinct) resize(newSizeDegree uint8) {
+	oldSize := p.bufSize()
+	oldBuf := p.buf
+
+	if 0 == newSizeDegree {
+		newSizeDegree = p.sizeDegree + 1
+	}
+
+	p.buf = make([]approxCountDistinctHashValue, uint32(1)<<newSizeDegree)
+	p.sizeDegree = newSizeDegree
+
+	// Move some items to new locations.
+	for i := uint32(0); i < oldSize; i++ {
+		x := oldBuf[i]
+		if x != 0 {
+			p.reinsertImpl(x)
+		}
+	}
+}
+
+func (p *partialResult4ApproxCountDistinct) readAndMerge(rb []byte) error {
+	rhsSkipDegree := rb[0]
+	rb = rb[1:]
+
+	if rhsSkipDegree > p.skipDegree {
+		p.skipDegree = rhsSkipDegree
+		p.rehash()
+	}
+
+	rb, rhsSize, err := codec.DecodeUvarint(rb)
+
+	if err != nil {
+		return err
+	}
+
+	if rhsSize > uint64(uniquesHashMaxSize) {
+		return errors.New("Cannot read partialResult4ApproxCountDistinct: too large size degree")
+	}
+
+	if p.bufSize() < uint32(rhsSize) {
+		newSizeDegree := max(uniquesHashSetInitialSizeDegree, uint8(math.Log2(float64(rhsSize-1)))+2)
```

Comment:
Should we consider the `lhsSize` for `newSizeDegree`? As `max(max(rhsSize+2, lhsSize+2), uniquesHashSetInitialSizeDegree)`.

Reply:
If `p.bufSize() >= uint32(rhsSize)`, there is no need to resize, the followed up process will insert element by `insertHash` and invoke `shrinkIfNeed`.

Reply:
OK, I got it.